File: AssistantPlus.Windows/View/ModdingWorker/ForwardHelper.cs
#pragma warning disable 0,
// ReSharper disable

using AssistantPlus;
using AssistantPlus.Utility;

namespace AssistantPlus.View.ModdingWorker {

	public static class ForwardHelper {

		#region utility

		public static async Task Forward (
			List<String> argument
		) {
			await App.MainWindow.InsertTabItem(new () {
				Title = ModuleHelper.Query(ModuleType.ModdingWorker).Name,
				Type = ModuleType.ModdingWorker,
				Option = ["-AdditionalArgument", ..argument],
			});
			return;
		}

		public static async Task ForwardMany (
			List<List<String>> argument,
			Boolean            parallel
		) {
			if (argument.Count != 0) {
				if (!parallel) {
					await ForwardHelper.Forward(argument.SelectMany(GF.ReturnSelf).ToList());
				}
				else {
					foreach (var argumentItem in argument) {
						await ForwardHelper.Forward(argumentItem);
					}
				}
			}
			return;
		}

		// ----------------

		public static List<String> MakeArgumentForCommand (
			String? input,
			String? method,
			Object? argument
		) {
			var command = new List<String>();
			command.Add(input ?? "?");
			if (method is not null) {
				command.Add("-method");
				command.Add(method);
			}
			if (argument is not null) {
				command.Add("-argument");
				command.Add(JsonHelper.SerializeText(argument, false));
			}
			return command;
		}

		#endregion

	}

}

File: module/assistant/lib/view/core_task_worker/forward_helper.dart
import '/common.dart';
import '/module.dart';
import '/setting.dart';
import '/utility/json_helper.dart';
import '/widget/export.dart';
import '/view/home/module_page.dart';
import '/view/core_task_worker/option.dart';

// ----------------

class ForwardHelper {

  // #region utility

  static Future<List<String>> execute(
    BuildContext context,
    List<String> argument,
  ) async {
    var setting = SettingProvider.of(context, listen: false);
    var key = GlobalKey();
    await setting.state.homeInsertPage!(
      key,
      .new(
        title: ModuleHelper.query(.coreTaskWorker).name,
        type: .coreTaskWorker,
        option: Option(
          immediateLaunch: false,
          additionalArgument: null,
        ),
      ),
      true,
    );
    var result = await key.currentState!.as<ModulePageState>().modulePageExecuteCommand('launch', argument);
    await setting.state.homeRemovePage!(key);
    return result;
  }

  // ----------------

  static Future<Void> forward(
    BuildContext context,
    List<String> argument,
  ) async {
    var setting = SettingProvider.of(context, listen: false);
    await setting.state.handleLaunch!(
      ModuleHelper.query(.coreTaskWorker).name,
      .coreTaskWorker,
      Option(
        immediateLaunch: true,
        additionalArgument: argument,
      ),
    );
    return;
  }

  static Future<Void> forwardMany(
    BuildContext       context,
    List<List<String>> argument,
    Boolean            parallel,
  ) async {
    if (!parallel) {
      await ForwardHelper.forward(context, argument.expand((value) => [...value]).toList());
    }
    else {
      for (var argumentItem in argument) {
        await ForwardHelper.forward(context, argumentItem);
      }
    }
    return;
  }

  // ----------------

  static String makeMethodMaybeBatch(
    String  method,
    Boolean batch,
  ) {
    return '${method}${!batch ? '' : '!batch'}';
  }

  static List<String> makeArgumentForCommand(
    String? input,
    String? method,
    Object? argument,
  ) {
    var command = <String>[];
    command.add(input ?? '?none');
    if (method != null) {
      command.add('-method');
      command.add(method);
    }
    if (argument != null) {
      command.add('-argument');
      command.add(JsonHelper.encodeText(argument, indented: false));
    }
    return command;
  }

  // #endregion

}
